### PR TITLE
Traits 6.0 updates

### LINKED ISSUE #769: Update for Traits 6.0
Where reasonable, update the codebase to use the new features of Traits 6.0.  This includes:

- replace `Either` with `Union`
- remove uses of `Trait()` where possible
- replace `Unicode` with `Str`
- replace other deprecated traits

### PR DESCRIPTION
**Pull request by corranwebster**
Make use of improvements from Traits 6.0.

Fixes #769

### CODE REVIEW

File: examples/demo/Extras/Tree_editor_with_TreeNodeRenderer.py [hunk 1]
```diff
@@ -14,17 +14,14 @@
 contents inside the cells of a TreeEditor.
 """
 
-    absolute_import, division, print_function, unicode_literals
```
Comment (mdickinson) [inline examples/demo/Extras/Tree_editor_with_TreeNodeRenderer.py]: Err. What happened here?

Comment (corranwebster) [reply examples/demo/Extras/Tree_editor_with_TreeNodeRenderer.py]: A bad search-replace for `from __future__` in #767

Comment (mdickinson) [reply examples/demo/Extras/Tree_editor_with_TreeNodeRenderer.py]: And the reviewer didn't catch it. For shame ...

File: examples/demo/Extras/Tree_editor_with_TreeNodeRenderer.py [hunk 2]
```diff
@@ -55,7 +52,7 @@ class MyData(HasTraits):
     """ The root node for a tree of data. """
 
     #: The name of the root node.
-    name = Unicode('Rooty McRootface')
+    name = Str('Rooty McRootface')
```
Comment (mdickinson) [inline examples/demo/Extras/Tree_editor_with_TreeNodeRenderer.py]: Good default node name!

File: traitsui/extras/demo.py [hunk 3]
```diff
@@ -113,7 +113,7 @@ def extract_docstring_from_source(source):
 
     Parameters
     ----------
-    source : Str (Unicode)
+    source : Str (Str)
```
Comment (mdickinson) [inline traitsui/extras/demo.py]: This one should probably just say "Str"

File: traitsui/qt4/code_editor.py [hunk 4]
```diff
@@ -20,7 +20,7 @@
 from pyface.ui.qt4.code_editor.code_widget import AdvancedCodeWidget
 from traits.api import (
     Str,
-    Unicode,
+    Str,
```
Comment (mdickinson) [inline traitsui/qt4/code_editor.py]: Redundant import

File: traitsui/undo.py [hunk 5]
```diff
@@ -72,11 +72,14 @@ class UndoItem(AbstractUndoItem):
...
 
     #: Object the change occurred on
-    object = Trait(HasTraits)
+    object = Instance(HasTraits)
+
     #: Name of the trait that changed
     name = Str
+
     #: Old value of the changed trait
     old_value = Property
```
Comment (mdickinson) [inline traitsui/undo.py]: change these ones from `Property` to `Property()` too? (Or do that in a separate PR?)

Comment (corranwebster) [reply traitsui/undo.py]: I might do that.  I added `()` to a lot of traits as part of the equivalent Pyface PR, but I can't remember how I did it reliably.  Probably just looking for the most common trait types.

Comment (corranwebster) [reply traitsui/undo.py]: On second thoughts, not in this PR - it touches too much code.
